enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+### Graduation Criteria
+
+#### Tech preview
+
+Operator handles OpenStack installation in 4.7.
+
+#### Tech Preview -> GA
+
+The operator will handle basic tasks, such as create initial templates at bootstrap and running all of supported in-tree providers after installation.
+
+## FAQ
+
+Q: Can we disaster-recover a cluster without CCM running?
+A: We can't. Basically, without CCM running, nodes can only join the cluster, but they will be unschedulable. Which means it's impossible to start any workloads on the nodes. [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)
+
+Q: What is non-functional in a cluster during bootstrapping until CCM is up?
+A: If CCM is not available, new nodes in the cluster will be left unschedulable.
+
+Q: Who talks to CCM?
+A: From an architectural point of view, only Kube API server is able to communicate with CCM. [Source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#design). All requests to CCM are sent by `kubelet`. KCM doesn't interact with CCM.
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration will be required for OpenStack and vSphere providers
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and communicates similarly to other kubernets components.
```

Comment:
So you do expose an endpoint, what it does?

Reply:
This information is wrong. This endpoint is used only for `healthz` and `readyz` and other services describing the component status, like metrics. All communication is done in the Node resource, no component is directly talking to the `CCM`. The out-of-tree CCM is vendoring the native cloud-provider code, no change from what `kube-controller-manager` did. It is just a default port, going to remove it.

Reply:
Right, but you are exposing an endpoint, even though it's for healthz and readyz it should be called out b/c it's an important piece of information for people looking at internal cluster communication.

Reply:
This has been updated in the last push and is now much more clear